x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_creation/components/configuration_step/configuration_step_form.tsx

```diff
@@ -304,17 +335,92 @@ export const ConfigurationStepForm: FC<CreateAnalyticsStepProps> = ({
+        }
+        return !option.key?.includes(runtimeMappingKey);
+      });
+      // Runtime mappings have been removed
+      if (runtimeMappings === undefined && runtimeMappingsUpdated === true) {
+        setDependentVariableOptions(filteredOptions);
+      } else if (runtimeMappings) {
+        // add to filteredOptions if it's the type supported
+        const runtimeOptions = getRuntimeDepVarOptions(jobType, runtimeMappings);
+        setDependentVariableOptions([...filteredOptions, ...runtimeOptions]);
+      }
+
+      if (resetDepVar) {
+        setIncludesTableItems([]);
+        setFormState({
+          dependentVariable: '',
+          includes: includes.filter((field) => {
+            const isRemovedRuntimeField = previousRuntimeMapping && previousRuntimeMapping[field];
+            return field !== dependentVariable || !isRemovedRuntimeField;
+          }),
+        });
+      } else if (hasBasicRequiredFields && hasRequiredAnalysisFields) {
+        const formCopy = cloneDeep(form);
+        // Remove old runtime fields from 'includes' table before sending to _explain
+        const filteredIncludes = includes.filter((field) => {
+          const isRemovedRuntimeField = previousRuntimeMapping && previousRuntimeMapping[field];
+          return !isRemovedRuntimeField;
+        });
+        formCopy.includes = filteredIncludes;
+        const { success, fieldSelection, errorMessage } = await fetchExplainData(formCopy);
+        if (success) {
+          // update the field selection table
+          const hasRequiredFields = fieldSelection.some(
+            (field) => field.is_included === true && field.is_required === false
+          );
+          setIncludesTableItems(fieldSelection);
+          setMaxDistinctValuesError(undefined);
+          setUnsupportedFieldsError(undefined);
+          setFormState({
+            includes: filteredIncludes,
+            requiredFieldsError: !hasRequiredFields ? requiredFieldsErrorText : undefined,
+          });
+        } else {
+          const {
+            maxDistinctValuesErrorMessage,
+            unsupportedFieldsErrorMessage,
+            toastNotificationDanger,
+            toastNotificationWarning,
+          } = handleExplainErrorMessage(errorMessage, sourceIndex, jobType);
+
+          if (toastNotificationDanger) {
+            toastNotifications.addDanger(toastNotificationDanger);
+          }
+          if (toastNotificationWarning) {
+            toastNotifications.addWarning(toastNotificationWarning);
+          }
 
-    const hasRequiredAnalysisFields =
-      (isJobTypeWithDepVar && dependentVariable !== '') ||
-      jobType === ANALYSIS_CONFIG_TYPE.OUTLIER_DETECTION;
+          setMaxDistinctValuesError(maxDistinctValuesErrorMessage);
+          setUnsupportedFieldsError(unsupportedFieldsErrorMessage);
+        }
+      }
+    }
+  }, [JSON.stringify(runtimeMappings)]);
+
+  useEffect(() => {
+    handleRuntimeUpdate();
+  }, [JSON.stringify(runtimeMappings)]);
```

Comment:
I assume you are missing a `debounce`.

Reply:
No debounce is needed here as this only runs when the user has clicked 'Apply change' when they've completed a runtime field - it's not running on every change.

Reply:
follow-up question: do we need a `JSON.stringify` if this object is updated only on the `Apply change` click? 🤔 can we just check the updated reference instead?